timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,299 @@
+          let timeRange = TimeRange (Just from) (fst <$> Map.lookupGT from values)
+           in f timeRange value
+      }
+    where
+      initialRange = TimeRange Nothing $ fst <$> Map.lookupMin values
+
+instance FoldableWithIndex TimeRange Timeline
+
+instance TraversableWithIndex TimeRange Timeline where
+  itraverse :: (Applicative f) => (TimeRange -> a -> f b) -> Timeline a -> f (Timeline b)
+  itraverse f = sequenceA . imap f
+
+changes :: Timeline a -> Set UTCTime
+changes Timeline {values} = Map.keysSet values
+
+data Record a = Record
+  { effectiveFrom :: UTCTime,
+    effectiveTo :: Maybe UTCTime,
+    value :: a
+  }
+  deriving stock (Eq, Functor)
+
+type Getter s a = forall f. (Contravariant f) => (a -> f a) -> s -> f s
+
+recordEffectiveFrom :: Getter (Record a) UTCTime
+recordEffectiveFrom f r = contramap getEffectiveFrom (f (getEffectiveFrom r))
+  where
+    getEffectiveFrom :: Record a -> UTCTime
+    getEffectiveFrom Record {effectiveFrom} = effectiveFrom
+
+recordEffectiveTo :: Getter (Record a) (Maybe UTCTime)
+recordEffectiveTo f r = contramap getEffectiveTo (f (getEffectiveTo r))
+  where
+    getEffectiveTo :: Record a -> Maybe UTCTime
+    getEffectiveTo Record {effectiveTo} = effectiveTo
+
+recordValue :: Getter (Record a) a
+recordValue f r = contramap getValue (f (getValue r))
+  where
+    getValue :: Record a -> a
+    getValue Record {value} = value
+
+-- | 'makeRecord' returns 'Nothing' if @effectiveTo@ is not greater than @effectiveFrom@
+makeRecord ::
+  -- | effective from
+  UTCTime ->
+  -- | optional effective to
+  Maybe UTCTime ->
+  -- | value
+  a ->
+  Maybe (Record a)
+-- can't write a makeRecordTH because UTCTime has no Lift instance
+makeRecord effectiveFrom effectiveTo value =
+  if maybe False (effectiveFrom >=) effectiveTo
+    then Nothing
+    else Just Record {..}
+
+instance (TH.Lift a) => TH.Lift (Record a) where
+  liftTyped Record {..} =
+    [||
+    Record
+      (unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime effectiveFrom))
+      (fmap unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime <$> effectiveTo))
+      $$(TH.liftTyped value)
+    ||]
+
+newtype LiftUTCTime = LiftUTCTime UTCTime
+  deriving stock (Generic)
+
+unLiftUTCTime :: LiftUTCTime -> UTCTime
+unLiftUTCTime (LiftUTCTime t) = t
+
+instance TH.Lift LiftUTCTime where
+  liftTyped (LiftUTCTime t) = [||LiftUTCTime (read $$(TH.liftTyped (show t)))||]
+
+instance (Show a) => Show (Record a) where
+  show Record {..} = show effectiveFrom <> " ~ " <> show effectiveTo <> ": " <> show value
+
+newtype Overlaps a = Overlaps {groups :: NonEmpty (OverlapGroup a)}
+  deriving newtype (Semigroup)
+  deriving stock (Eq, Generic)
+
+instance (Show a) => Show (Overlaps a) where
+  show Overlaps {groups} =
+    "Here are " <> show (length groups) <> " group(s) of overlapping records\n"
+      ++ sep
+      ++ intercalate sep (show <$> NonEmpty.toList groups)
+      ++ sep
+    where
+      sep = "--------------------\n"
```

Comment:
I object to this being an instance of `show`

Reply:
I'll change them to `prettyXXX`